Otus-DevOps-2018-02/ifqthenp_microservices
# HW 15. Docker images & microservices (docker-3)

# Completed Homework №15

- :large_blue_diamond: [x] Main tasks
- :large_orange_diamond: [x] Extra tasks for self-study

## The following tasks have been completed

- :large_blue_diamond: Added three components to the `src` directory for Reddit application microservices
  - `post-py`
  - `comment`
  - `ui`
- :large_blue_diamond: Created three Dockerfiles for the microservices and optimized Dockerfile commands to reduce image sizes
- :large_blue_diamond: Created three Docker images with `docker build` command
- :large_blue_diamond: Created bridged network `reddit`, assigned network aliases for containers, and launched containers in `reddit` network
- :large_orange_diamond: Stopped running containers and relaunched them with new network aliases and environment variables:
  1. `docker run -d --network=reddit --network-alias=post_db_new --network-alias=comment_db_new mongo:latest`
  2. `docker run -d --network=reddit --network-alias=post_new -e POST_DATABASE_HOST='post_db_new' -e POST_DATABASE='posts_new' ifqthenp/otus-reddit/post:1.0`
  3. `docker run -d --network=reddit --network-alias=comment_new -e COMMENT_DATABASE_HOST='comment_db_new' -e COMMENT_DATABASE='comments_new' ifqthenp/otus-reddit/comment:1.0`
  4. `docker run -d --network=reddit -p 9292:9292 -e POST_SERVICE_HOST='post_new' -e COMMENT_SERVICE_HOST='comment_new' ifqthenp/otus-reddit/ui:1.0`
- :large_orange_diamond: Created new images based on Linux Alpine and applied best practices to reduce image sizes for `post-py`, `comment`, and `ui` microservices. New pull request has been created on separate branch `docker-3-alpine`

## How to get project running

1. Connect to the previously created `docker-host` on GCP
    - `eval $(docker-machine env docker-host)`
2. `cd` into `src` folder and build Docker images
    - `docker pull mongo:latest`
    - `docker build -t ifqthenp/otus-reddit/post:1.0 ./post-py`
    - `docker build -t ifqthenp/otus-reddit/comment:1.0 ./comment`
    - `docker build -t ifqthenp/otus-reddit/ui:2.0 ./ui`
3. Create Docker network for Reddit application
    - `docker network create reddit`
4. Create volume for MongoDB
    - `docker volume create reddit_db`
5. Run containers 
    - `docker run -d --network=reddit --network-alias=post_db --network-alias=comment_db -v reddit_db:/data/db mongo:latest`
    - `docker run -d --network=reddit --network-alias=post ifqthenp/otus-reddit/post:1.0`
    - `docker run -d --network=reddit --network-alias=comment ifqthenp/otus-reddit/comment:1.0`
    - `docker run -d --network=reddit -p 9292:9292 ifqthenp/otus-reddit/ui:2.0`
6. Kill running containers
    - `docker kill $(docker ps -q)`

## How to check that project actually works

- Ensure that the build passed TravisCI tests
  - [#andrei_bogomja](https://devops-team-otus.slack.com/messages/C9M6L1XB3) Slack channel
  - actual GitHub pull request page
- Type URL in browser address bar `http://35.197.227.48:9292/` and post a link or leave a comment
- Kill running containers and relaunch them again to make sure that the data has not been lost

## PR checklist
